### PR TITLE
Added link support to all templates

### PR DESCRIPTION
**Pull request by thegecko**
It's a shame the left/right links aren't supported on styles apart from the social one.

This PR adds support to the other styles.

### CODE REVIEW

**Comment by thegecko** [comment]
Looks like the tests failed as the older templates were expected. If this is a valuable PR, I'll update the tests, too.

**Comment by ELLIOTTCABLE** [comment]
I'm curious why this hasn't been merged yet; I'd really like to see this added.

Specifically, I'm [using Shields.io](https://gist.github.com/ELLIOTTCABLE/a9f9a7e6264130d08d76) to replace <StillMaintained.com> in my projects, with badges like the following:



I'd really like to be able to able to link the word ‘feature frozen’ to a README section below, with something like `?link=%23feature-freeze`.

**Comment by espadrine** [comment]
@thegecko The tests fail. Uncoincidentally, the template fails to parse. I am not sure why, but it seems this patch as is cannot work.

File: templates/flat-template.svg [hunk 1]
```diff
@@ -19,8 +19,20 @@
       <image x="5" y="3" width="{{=it.logoWidth}}" height="14" xlink:href="{{=it.logo}}"/>
     {{?}}
     <text x="{{=(it.widths[0]+it.logoWidth+it.logoPadding)/2}}" y="15" fill="#010101" fill-opacity=".3">{{=it.escapeXml(it.text[0])}}</text>
+    {{?it.links[0]}}
+    <a xlink:href="{{=it.links[0]}}">
+      <text x="{{=(it.widths[0]+it.logoWidth+it.logoPadding)/2}}" y="14" fill="#fff">{{=it.escapeXml(it.text[0])}}</text>
+    </a>
+    {{??}}
     <text x="{{=(it.widths[0]+it.logoWidth+it.logoPadding)/2}}" y="14">{{=it.escapeXml(it.text[0])}}</text>
+    {?}
```
Comment (espadrine) [inline templates/flat-template.svg]: There should be two: `{{?}}`.

File: templates/plastic-template.svg [hunk 2]
```diff
@@ -18,8 +18,20 @@
 
   <g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="11">
     <text x="{{=it.widths[0]/2+1}}" y="14" fill="#010101" fill-opacity=".3">{{=it.escapeXml(it.text[0])}}</text>
+    {{?it.links[0]}}
+    <a xlink:href="{{=it.links[0]}}">
+      <text x="{{=it.widths[0]/2+1}}" y="13" fill="#fff">{{=it.escapeXml(it.text[0])}}</text>
+    </a>
+    {{??}}
     <text x="{{=it.widths[0]/2+1}}" y="13">{{=it.escapeXml(it.text[0])}}</text>
+    {?}
```
Comment (espadrine) [inline templates/plastic-template.svg]: `{{?}}`

File: templates/plastic-template.svg [hunk 3]
```diff
@@ -18,8 +18,20 @@
...
     <text x="{{=it.widths[0]+it.widths[1]/2-1}}" y="14" fill="#010101" fill-opacity=".3">{{=it.escapeXml(it.text[1])}}</text>
+    {{?it.links[1]}}
+    <a xlink:href="{{=it.links[1]}}">
+      <text x="{{=it.widths[0]+it.widths[1]/2-1}}" y="13" fill="#fff">{{=it.escapeXml(it.text[1])}}</text>
+    </a>
+    {{??}}
     <text x="{{=it.widths[0]+it.widths[1]/2-1}}" y="13">{{=it.escapeXml(it.text[1])}}</text>
+    {?}
```
Comment (espadrine) [inline templates/plastic-template.svg]: {{?}}

File: templates/flat-template.svg [hunk 4]
```diff
@@ -19,8 +19,20 @@
...
     <text x="{{=it.widths[0]+it.widths[1]/2-1}}" y="15" fill="#010101" fill-opacity=".3">{{=it.escapeXml(it.text[1])}}</text>
+    {{?it.links[1]}}
+    <a xlink:href="{{=it.links[1]}}">
+      <text x="{{=it.widths[0]+it.widths[1]/2-1}}" y="14" fill="#fff">{{=it.escapeXml(it.text[1])}}</text>
+    </a>
+    {{??}}
     <text x="{{=it.widths[0]+it.widths[1]/2-1}}" y="14">{{=it.escapeXml(it.text[1])}}</text>
+    {?}
```
Comment (espadrine) [inline templates/flat-template.svg]: two {{}}

File: templates/flat-square-template.svg [hunk 5]
```diff
@@ -1,10 +1,22 @@
-<svg xmlns="http://www.w3.org/2000/svg" width="{{=it.widths[0]+it.widths[1]}}" height="20">
+<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="{{=it.widths[0]+it.widths[1]}}" height="20">
   <g shape-rendering="crispEdges">
     <rect width="{{=it.widths[0]}}" height="20" fill="{{=it.escapeXml(it.colorA||"#555")}}"/>
     <rect x="{{=it.widths[0]}}" width="{{=it.widths[1]}}" height="20" fill="{{=it.escapeXml(it.colorB||"#4c1")}}"/>
   </g>
   <g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="11">
+    {{?it.links[0]}}
+    <a xlink:href="{{=it.links[0]}}">
+      <text x="{{=it.widths[0]/2+1}}" y="14" fill="#fff">{{=it.escapeXml(it.text[0])}}</text>
```
Comment (espadrine) [inline templates/flat-square-template.svg]: Why `fill="#fff"`?

**Comment by espadrine** [comment]
Could you please indent the contents of `{{?…}}…{{?}}` please? It would gain in readability.

File: templates/flat-square-template.svg (hunk 5 above)
Comment (thegecko) [inline templates/flat-square-template.svg]: This is the colour used in the social template, what should it be?

**Comment by thegecko** [comment]
@espadrine This looks better

**Comment by espadrine** [comment]
Chrome has a display bug for your code:
chrome-shields

The bug is filed at https://code.google.com/p/chromium/issues/detail?id=352912. It is seemingly inconsistent with Firefox and Safari.

As a result, the only way to get cross-platform results is to put `fill=#fff`, which explains this: https://github.com/badges/shields/pull/598#discussion_r50480144.

**Comment by thegecko** [comment]
Do you want me to add the fill=#fff back in to match the social template or will it fail the tests? Perhaps The dirty fix should be implemented for all templates including the social one?

**Comment by espadrine** [comment]
It would still be erased by SVGo. We simply need Google to fix their software, and when it's been released in the wild for a month, I'll merge this.

**Comment by olivierlacan** [comment]
@thegecko @espadrine Any movement on SVGo or the current merge conflict?

**Comment by paulmelnikow** [comment]
As this can't be merged and the problem persists, let's track this via #1079.

@espadrine has rebased this work onto master in [this working branch](https://github.com/espadrine/shields/tree/link).

**Pull request closed without merging**